docs/guides/magicgui.md

```diff
@@ -0,0 +1,530 @@
+{class}`~napari.layers.Points`), it will be rendered as a
+{class}`~magicgui.widgets.ComboBox` widget (i.e. "dropdown menu"), where the
+options in the dropdown box are the layers of the corresponding type currently
+in the viewer.
+
+```python
+from napari.layers import Image
+
+@magicgui
+def my_widget(image: Image):
+    # do something with whatever image layer the user has selected
+    # note: it *may* be None! so your function should handle the null case
+    ...
+```
+
+Here's a complete example:
+
+```{code-cell} python
+:tags: [remove-output]
+import napari
+import numpy as np
+from napari.layers import Image
+
+@magicgui(image={'label': 'Pick an Image'})
+def my_widget(image: Image):
+    ...
+
+viewer = napari.view_image(np.random.rand(64, 64), name="My Image")
+viewer.window.add_dock_widget(my_widget)
+```
+
+*Note the widget at the bottom with "My Image" as the currently selected option*
+
+```{code-cell} python
+:tags: [remove-input]
+from napari.utils import nbscreenshot
+
+viewer.window._qt_window.resize(750, 550)
+nbscreenshot(viewer)
+```
+
+### Annotating as `Layer`
+
+In the previous example, the dropdown menu will *only* show
+{class}`~napari.layers.Image` layers, because the parameter was annotated as an
+{class}`~napari.layers.Image`.  If you'd like a dropdown menu that allows the
+user to pick from *all* layers in the layer list, annotate your parameter as
+{class}`~napari.layers.Layer`
+
+```python
+from napari.layers import Layer
+
+@magicgui
+def my_widget(layer: Layer):
+    # do something with whatever layer the user has selected
+    # note: it *may* be None! so your function should handle the null case
+    ...
+```
+
+### Annotating as `napari.types.*Data`
+
+In the previous example, the object passed to your function will be the actual
+{class}`~napari.layers.Layer` instance, meaning you will need to access any
+attributes (like `layer.data`) on your own.  If your function is designed to
+accept a numpy array, you can use the any of the special `<LayerType>Data` types
+from {mod}`napari.types` to indicate that you only want the data attribute from
+the layer (where `<LayerType>` is one of the available layer types).  Here's an
+example using {attr}`napari.types.ImageData`
+
+```python
+from napari.types import ImageData
+import numpy as np
+
+@magicgui
+def my_widget(array: ImageData):
+    # note: it *may* be None! so your function should handle the null case
+    if array is not None:
+      assert isinstance(array, np.ndarray)  # it will be!
+```
+
+### Annotating as `napari.Viewer`
+
+Lastly, if you need to access the actual {class}`~napari.viewer.Viewer` instance
+in which the widget is docked, you can annotate one of your parameters as a
+{class}`napari.Viewer`.
+
+```python
+from napari import Viewer
+
+@magicgui
+def my_widget(viewer: Viewer):
+  ...
+```
+
+```{caution}
+Please use this sparingly, as a last resort. If you need to *add* layers
```

Comment:
Can we give an example where it would be reasonable to provide the Viewer? Are there such examples?

Reply:
I hesitate to give an example, because of how frequently people see an example and just copy/paste it assuming it's best practice.  But, this would be necessary if you needed one of the viewer attributes that didn't exist on the layers themselves (like theme, ndisplay, grid, perhaps viewer.status, etc...).  Again, it's nothing we want to promote as a "go to" API, but it's there because people have needed it.  See for example, [usage in affinder](https://github.com/jni/affinder/blob/b538689a92ce878efae8316cb981fd0c9fb9ced8/affinder/affinder.py#L96).

Reply:
Yeah, I was wondering about this bit too. We can probably leave for now, but we might want to update as we get more experience with when this is necessary etc.

Reply:
or when we can suggest other alternatives etc.